\documentclass[10pt,a4paper]{article}
\usepackage[latin1]{inputenc}
\usepackage[top=1in, bottom=1.25in, left=1.25in, right=1.25in]{geometry}
\usepackage{amsmath}
\usepackage{amsfonts}
\usepackage{amssymb}
\usepackage{graphicx}
\author{benjamin Peter}

\newcommand{\BE}[1]{\mathbb{E}\left[#1\right]}

\title{Admixfrog model}
\begin{document}
	\maketitle
\section*{Introduction}

	We aim to build an HMM that classifies a single \emph{target} genome into regions that were derived from different diverged populations. This is motivated by admixture between modern humans, Neandertals or Denisovans, (which I will occasionally refer to as source populations) although the model is applicable to many other taxa. 
	
	Briefly, the hidden states are the sources of the individual, e.g. homozygous modern human (HH) homozygous Denisovan (DD) and homozygous Neandertal (NN), as well as the heterozygosu states HD, DN, HN where one haplotype each is derived from the sources.
	
	We assume a homogeneous Markov model throughout, which we achieve by binning the genome in bins  of constant map size (e.g. 0.005 cM), so it is possible that a bin has zero, one or many emissions, which are treated independently distributed given the hidden state. We assume that data comes in the form of biallelic SNP. We assume we have a sample of genotypes from each source, as well as reads covering the SNP from the target.
	
	Emissions are the number of reads of the ancestral /derived allele, with the allele frequency in Humans / Neandertals / Denisovans assumed known.
	
	The transition matrix is estimated using standard- Baum-Welch. The rest of this note deals with inferring the emission probabilities, and in particular estimating contamination.
\section*{Notation}
	Let 
	\begin{itemize}
		\item $R, L$ the number of read groups, and hidden-state bins, respectively
		\item $S_l$ the number of SNP in bin $l$
		\item $n_{rs}$ the number of reads of readgroup $r$ at SNP $s$
		\item $O = (O_{rs})$ the set of derived-allele- read counts from library $r$ at SNP $s$
		\item $Z = (Z_l)$ be the sequence of hidden states
		\item $R = (R_{l})$ the number of contaminant reads at SNP $l$
		\item $A_{sk}, B_{sk}$ the number of derived and ancestral allele counts in reference $k$ at SNP $s$, (or more generally parameters of a Beta prior)
		\item $T$ be a transition matrix		
		\item $F_k$ a measure of distance between source $k$ and the introgressing population
		\item $c_r$ proportion of contaminant reads in read group $r$
		\item $p_{sc}$ the allele frequency of a contaminant at position $s$. 
		\item $e$ the sequencing error rate
		\item $\mathbf{\theta} = (F_k, c_r, T, e)$, the set of all parameters to be estimated
	\end{itemize}
The model can be summarized as
	\begin{itemize}
		\item $O_{rs} | G_{s}, n_{rs} \sim Binomial(O_{rs}; n_{rs}, p(e, c_r, p_{sc}, G_{s}))$
		\item $G_s|Z_l=k \sim Binomial(2, f_{sk})$
		\item $f_{s} | Z_l=k \sim Beta(a_kF_k, b_kF_k)$
		\item $Z_s | Z_{s-1}= k' \sim T_{k'}$
		\item $Z_0 \sim \alpha_0$
	\end{itemize}
	Where we want to estimate the $c_r, F_k, T$ and $\alpha_0$

\section{Algorithm details}
We factor the likelihood as 
$$P(O | \theta) =  \sum_{\mathcal{G}, \mathcal{Z}} P(O | G, c, p_c, N)P(G | Z, A, B) P(Z| T)$$
Where $\mathcal{G}, \mathcal{Z}$ are the sets of all possible genotype assignments.

Under the Markov assumption, this can be further split up as
$$P(O | \theta) = P(Z_0) +
\sum_{k=1}^K \sum_{g=0}^2 \sum_{l=1}^L 
P(O_l | G_l, c, p_{cn}, N_l)
P(G_l | Z_l = k, A_{lk}, B_{lk}, \tau)
P(Z_l = k |Z_{l-1}, T) $$
we refer to these terms as the genotype likelihood, state likelihood and transition probability, respectively.

A standard trick is to assume the latent variables $G, Z$ are known. Taking the log
yields the complete data likelihood
\begin{equation}
\mathcal{L}  = P(O, Z, G | \theta) = \log P(Z_0) 
+ \sum_{l=1}^L \log P(O_l | G_l, c)
+ \sum_{l=1}^L \log P(G_l | Z_l = k, F)
+ \sum_{l=1}^L  \log P(Z_l |Z_{l-1}, T) \label{eq:ll:simple}
\end{equation}

The model is further improved by taking into account the each bin has $S_L$ snps, and we have $R$ distinct read groups.
\begin{align}
\mathcal{L}  &= P(O, Z, G | \theta)  \nonumber\\
&=\log P(Z_0) \nonumber\\
&+ \sum_{r=1}^R\sum_{s=1}^{S_L}\sum_{l=1}^L \log P(O_r | G_s, c)\nonumber\\
&+ \sum_{s=1}^{S_L}\sum_{l=1}^L \log P(G_s | Z_l = k, F)\nonumber\\
&+ \sum_{l=1}^L  \log P(Z_l |Z_{l-1}, T) \label{eq:ll:correct}
\end{align}

The parameters we want to estimate are the contamination rate $c_r$ for each read group, the $F_{ST}$ for each source population, and the sequence of hidden states $Z$.

\subsection{Genotype Likelihood}
\begin{equation}
P(O_{rs}| G_s, c_r, n_{rs}) \sim Binom(O_{rs}; n_{rs}, p) \label{eq:ll:geno}
\end{equation}
where $p = (1-e) p' + e (1-p')$
and $p' =c_{rs} p_c + (1-c_{rs}) G_s$
\subsection{State Likelihood}
We want to calculate the probability of a single locus, given the hidden state (i.e ancestry) is known. For homozygous ancestry, if the allele frequency of the introgressed individual were known to be $p$, the genotype probabilities would be $p, 2p(1-p), (1-p)^2$, respectively However, we do not observe $p$, but instead need  to estimate it from a possibly distant proxy. There are three causes of uncertainty that we will take into account.
\begin{enumerate}
	\item Sampling uncertainty from the reference.	
	\item Genetic drift between the reference and introgressing population
	\item Population substructure in the reference population
\end{enumerate}


\subsubsection{Homozygous Hidden States}
As we deal with a single hidden-state bin and locus, subscripts designating the genomic location are omitted in this section. We start with the simplest possible model, and add complexity in sequence.
\paragraph{Direct samples}
We wish to calculate the probability that the individual is homozygous, given the SNP of interest derives two haplotypes from a population whose allele frequency is known to be $p$. If, $p$ is not known, 
\begin{equation}
G \sim Binom(2, p)
\end{equation}
however, we may need to infer it from a panel where we observe $a', d'$ ancestral and derived alleles, respectively. In this case
\begin{align}
p &\sim Beta(d' + d_0, a' + a_0)
\end{align}
The prior $d_0, a_0$ can be justified by noting that even if we only observe ancestral alleles in our sample, there is still a small chance that some individuals might have a derived allele in this locus. There are differing opinions about the values of $d_0, a_0$. Balding \& Nichols proposed a uniform prior, i.e. $a_0=d_0=1$. By noting that the derived allele occurs proportionally to $\theta/p$, one can justify $d_0=0, a_0=1$. A third option is an empirical Bayes prior: In the absence of data at a particular locus, a sensible guess for the allele frequency at this locus would be the genome wide average. Hence, we obtain an empirical prior by fitting a $Beta(d_0, a_0)$ distribution to all loci in the reference panel. In particular for ascertained data, this might be most suitable.

Writing $d = d' + d_0, a = a'+a_0$
\begin{eqnarray}
P(G | d, a, d_0, a_0) &\sim& Betabinom(2, d+d_0, a+a_0)\nonumber\\
&=& \binom{2}{G} \frac{B[G+d, 2-G + a]}{B[ d + a]} \label{eq:ll:homo}
\end{eqnarray}
where $B[.]$ denotes the beta function.
\paragraph{Genetic drift between reference and target}
In many cases, the target individual will not be sampled immediately after admixture, but thousands of generations after. Thus, we need to take into account that the allele frequency might have changed.

Conceptually, genetic drift will increase the chance that the two alleles in the target are derived from the same ancestor: If the target is very very distant from the panel, the probability $F$ that the two alleles are identical by descent approaches one.  Conversely, if we have a very close reference, that probability is zero. 
\begin{align}
P(G=2 | a,d, F) &= \int P(G|p) P(p | a, d) dp \nonumber\\
&= \int \left[ F p + (1-F) p^2\right] P(p | a, d) df\nonumber\\
&= F \BE{p|a, d} + (1-F)  \BE{p^2|a, d}\nonumber\\
&= F \frac{d}{a+d} + (1-F)  \frac{d (d+1)}{(a+d)(a+d+1)}\nonumber\\
&= \frac{d^2 + d + adF}{(a+d)(a+d+1)}
\end{align}
Hence we have an excess of homozygous emissions. The probability for $G=0$ is obtained by switching $A$ and $D$, and for $G=1$ by subtracting them from one:
\begin{equation}
P(G=1 | a,d) = \frac{2ad(1-F)}{(a+d)(a+d+1)}
\end{equation} 

\paragraph{Drift to reference, population structure in source}
As a final complication, the reference may be subdivided. For example, we want to infer Neandertal introgression by using the Altai Neandertal \\citep{prufer2014} as a reference, despite it being substantially diverged. This will lead to additional uncertainty in the allele frequency, but should not change our belief in the average allele frequency. For this purpose, we introduce a parameter $\tau$

\begin{equation}
p| a, d, \tau \sim Beta (d\tau, a\tau)
\end{equation}



\begin{align}
P(G=2 | a,d, F, \tau) &= \int P(G|p) P(p | a, d, \tau) df\nonumber\\
&= \int \left[ F p + (1-F) p^2\right] P(p | a, d, \tau) df\nonumber\\
&= F E[p|a,d] + (1-F)  E[p^2|a, d, \tau]\nonumber\\
&= F \frac{d}{a+d} + (1-F)  \frac{d (d\tau+1)}{(a+d)(a\tau+d\tau+1)}\nonumber\\
&= \frac{\tau d^2 + d + adF\tau}{(a+d)(\tau a+\tau d+1)}
\end{align}
The probability for $G=0$ is again obtained by switching $a$ and $d$, and for $G=1$ by subtracting them from one:
\begin{equation}
P(G=1 | a,d, F, \tau) = \frac{2ad(1-F)\tau}{(a+d)(\tau a+\tau d+1)}
\end{equation} 

In the limit as $\tau=1$, we have the same as without subdivision. In the 
other limit as $\tau=0$, we get
\begin{equation}
P(G=2|a, d, F, \tau) = d / (a+d)
\end{equation}

\paragraph{Drift to reference / subdivided source / alt}
A more complex model is
\begin{equation}
p| a, d, \tau \sim Beta (d'\tau + d_0, a'\tau + a_0)
\end{equation}




\subsubsection{Heterozygous States}
For a heterozygous state, we know the allele frequencies in the two source pops, and we know they contribute one haplotype each. If we assume that the sources have allele frequencies $p_i$ with prior $d_i, a_i$, respectively.

\begin{align}
P(G_i = 1 | \tau_i, F_i, d_i, a_i) &=  \int p P(p| d_i, a_i, p \tau_i, F_i) df\nonumber\\
&= E[f| d_i, a_i, F_i, \tau_i]\nonumber\\
&=\frac{d_i}{a_i+d_i}\label{eq:ll:het}
\end{align}
and hence is independent from $F$ and $\tau$. Intuitively, this is because if we just sample one allele from a population, that population's allele frequency is the best guess for the state of our sample. The distance increases our uncertainty, but not the mean.
Thus for sources $S_i, S_j$, 
\begin{align*}
P(G=0) &= \frac{a_i a_j}{(d_i+a_i)(d_j+a_j)}\\
P(G=1) &= \frac{a_id_j + a_jd_i}{(d_i+a_i)(d_j+a_j)}\\
P(G=2) &= \frac{d_i d_j}{(d_i+a_i)(d_j+a_j)}\\
\end{align*}



\subsubsection{Haploid States}
On the sex chromosome, or when inbreeding is present, we further encounter haploid regions. Here, the heterozygous states have emission probability 0, and we just have one allele of the heterozygous case: $F_{ST}$ does not matter.
\begin{align}
P(G=0) &= \frac{a}{a+d}\nonumber\\
P(G=1) &= \frac{d}{a+d}\label{eq:ll:hap}
\end{align}




\subsection{Estimating F and $\tau$}
\begin{align}
Q(F, \tau|F', \tau' )&= E[\log P(O, Z, G )  P(G|Z, \theta') P(Z | \theta')]\nonumber\\
&=\sum_{Z \in \mathcal{Z}, G \in \mathcal{G}} \log P(G  | Z, A, B, \tau) P(G|Z, \theta') P(Z | O, \theta')\nonumber\\
&=\sum_k\sum_{g=0}^2\sum_{l=1}^L \log P(G_t | Z_l=k, A_{lk}, B_{lk}, F)  P(G_l|Z_l, F', O) P(Z_l=k | O, \theta')
\end{align}
Since the $F_k$-parameters are independent for each homozygous state, these likelihoods are optimized numerically and independently.

\begin{equation}
\hat{F}_k =\operatorname*{argmax}_F \left[ \sum_{g=0}^2\sum_{l=1}^L \log P(G_t | Z_l=k, A_{lk}, B_{lk}, F)  P(G_l|Z_l, F', O) P(Z_l=k | O, \theta') 
\right]\label{eq:opt:F}
\end{equation}

where $P(Z_l = k |O, \theta')$ is the output of the forward-backward algorithm, $P(G_s | Z_l=k, A_{sk}, B_{sk}, F_k)$ is given by (\ref{eq:ll:homo}).   and
\begin{equation}
P(G_i | Z_i=k, O) = \frac{P(O_i | G_i, c') P(G_i | Z_i=k, F')}{ \sum_{g=0}^2 P(O_i | G_i=g, c') P(G_i=g | Z=k, F_k') }
\label{eq:em:G}
\end{equation}
This equation follows from 
\begin{align}
P(G,Z| O) &= P(G|Z, O)P(Z|O)\nonumber\\
&=P(G | Z, O_i) P(Z | O)\nonumber\\
&= \frac{P(O_i | G) P(G|Z)}{P(O_i | Z)} P(Z|O)
\end{align}
as $G_i | Z_i$ is independent of all observations except $O_i$ and Bayes theorem.




\subsection{Estimating Contamination}
\begin{align}
Q(c|c' )&= E[\log P(O, Z, G )  P(G|Z, \theta') P(Z | \theta')]\nonumber\\
&=\sum_{Z \in \mathcal{Z}}\sum_{G \in \mathcal{G}} \log P( O | G, c) P(G|Z, \theta') P(Z | O, \theta')\nonumber\\
&=\sum_{r=1}^R\sum_k\sum_{g=0}^2 \sum_{s=1}^{S_l}\sum_{l=1}^L \log P(O_{lr} | G_s=g, c_r)  P(G_s|Z_l=k, \theta', O) P(Z_l=k | O, \theta')
\end{align}

We thus optimize
\begin{align}
\hat{c}_r = \operatorname*{argmax}_{c_r}\sum_k\sum_{g=0}^2 \sum_{s=1}^{S_l}\sum_{l=1}^L \log P(O_{lr} | G_s=g, c_r)  P(G_s|Z_l=k, \theta', O) P(Z_l=k | O, \theta')
\end{align}
using (\ref{eq:em:G}) and (\ref{eq:ll:geno}).
\subsection{Estimationg Transitions and Initial State}
$T$ is estimated using standard Baum-welch, $\alpha_0$ is set to the stationary distribution of $T$.
\subsubsection{Transitions(HWE)}
$$\hat{T}_{ij} = \frac{\sum_l P(X_l=i, X_{l+1}=j| \theta', O)}{ P(O | \theta')}$$
However, the two haplotypes might change ancestry independently from each other. If the transitions from haploid states $H_{ab}$ is known, then 
$$T_i(j) = T_{i_1, i_2}(j_1, j_2) = \frac{H_{i_1}(j_1)H_{i_2}(j_2) + H_{i_1}(j_2)H_{i_2}(j_1)}{2}$$
i.e. there are always two possible transitions. We can introduce another inbreeding parameter $f$, so that when $i_1=i_2, j_1=j_2$ 
$T_i(j) = f H_i(j) + (1-f) H_i(j)^2$. Conversely, if there is only a single introgressed lineage, we would only ever expect haploid introgressed segments:



Compared to standard BW
$$\xi_{ij}(t) \propto P(O_{[1\dots l]}, Z_l | \theta') T_{ij}  P(O_{[l+2\dots L]}|Z_{l+1}, \theta') P(O_{l+1}|Z_{l+1}=k)$$

\begin{align}
P(H_l = N, H_{l+1} = A, O |\theta) &=  P(O_{[1\dots l]}, H_l | \theta') P(H_{l+1} | H_l) P(O_{[l+2\dots L]}|H_{l+1}, \theta') P(O_{l+1}|H_{l+1})\\
&=P(AN \to NN) + P(AA \to AN) + P(AA \to NN)
\end{align}

\section{Prior}
There are a few options for priors. For computational reasons, a beta prior is most suitable.
One possibility is to use a $Beta[0,0]$ or $Beta[1,1]$, which can be done by adding psuedocounts to the observed data.
More interesting is the assumption that the observed alleles were drawn from a neutral population with allele frequencies proportional to 
\begin{equation}
P(f) \propto \frac{\theta}{x}
\end{equation}
If we encounter a sample with $a$ ancestral and $d$ derived alleles, we have by Bayes' theorem
\begin{align}
P(f | a, d) &\propto P(f) P(a, d|f) \nonumber\\
 &\propto \frac{\theta}{f} \binom{a+d}{d} f^d (1-f)^a \nonumber\\
 &\propto f^{d-1} (1-f)^a \nonumber\\
&\sim Beta(d, a+1)
\end{align}
I.e. knowing the ancestral allele is equivalent to adding an additional ancestral allele to the observation. If the ancestral allele is unknown, ``mirroring this distribution``

For the folded SFS,
\begin{align}
P(f | a, d) &\propto P(f) P(a, d|f) \nonumber\\
&\propto \left[ \frac{1}{f} + \frac{1}{1-f}\right] f^d (1-f)^a\nonumber\\
&\propto \frac{1}{f(1-f)} f^d (1-f)^a\nonumber\\
&\propto f^{d-1} (1-f)^{a-1} \nonumber\\
&\sim Beta(d, a)
\end{align}

In a finite population, alleles enter a population at a frequency of $\frac{1}{2N}$, and will become fixed if at frequency $1-\frac{1}{2N}$. Writing
$$\eta(N) = \frac{1}{2N \sum_{i=1}^{2N-1}1/i}$$
We find that the shape of the SFS is well approximated by 
\begin{align}
P(f | a, d, N) &\sim Beta(d + \eta, a+1 - \eta)\\
&\sim Beta(d + \eta, a + \eta)
\end{align}
for the unfolded and folded case, respectively. As the frequency cannot exceed the limits, we need to truncate this distribution:
\begin{align}
E(f | a, d, N) &= \frac{B[\frac{1}{2N}, a+1, d]-B[1-\frac{1}{2N}, a+1, d]}{B[\frac{1}{2N}, a, d]-B[1-\frac{1}{2N}, a, d]}\nonumber\\
&= \frac{a}{a+d} \frac{\int_\epsilon^{1-\epsilon} t^a(1-t)^{d-1}dt}
{\int_\epsilon^{1-\epsilon} t^{a-1}(1-t)^{d-1}dt}\nonumber\\
&= V \frac{a}{a+d}\\
E(1-f | a, d, N)&= 1-V\frac{a}{a+d} = \left[1 + \frac{a}{d} (1-V)\right] \frac{d}{a+d}\\
\end{align}
The term in the squared parentheses is also denoted as $V'$. $V$ in general will depend on 
		
\subsection{Track lengths}
To a first approximation, the length of introgressed fragment is 
$L \sim Exp[r t ]$, where $r$ is a recombination rate and $t$ is the time since the fragment introgressed. However, tracks may overlap or be adjacent in a diploid genome. A simple model to correct for this is to assume that after a track finishes, with some probability $p$ a new introgressed track with the same distribution starts. Let us call this probability $m$. Then

\begin{equation}
L \sim Exp[(1-m)rt]
\end{equation}
writing $K = \frac{1-is}{rt}$, the characteristic function of $L$ is
\begin{align}
C(s) &= E[\exp(si [(1-m)L + (1-m)m (L+L')+ \dots  ] )]\nonumber\\
&=(1-m) K^{-1} + (1-m)m K^{-2} +(1-m)m^2 K^{-3} + \dots\nonumber\\
&=\frac{(1-m)}{K} \sum_{i=0}^\infty \left(\frac{m}{K}\right)^{i} \nonumber\\
&= \frac{(1-m)}{K} \frac{1}{1 - m/K}\nonumber\\
&= \frac{1-m}{K-m}\nonumber\\
&= \frac{rt(1-m)}{rt (1-m) - is}\nonumber
\end{align}
which is the characteristic function of an exponential distribution with parameter $rt(1-m)$.

\end{document}